#
# This patch contains a couple of PAM enhancements:
#   1) Each SSHv2 userauth method has its own PAM service name so that PAM can
#      be used to control what userauth methods are allowed.
#   2) The PAMServiceName and PAMServicePrefix options.
# 
# We have contributed back this feature to the OpenSSH upstream community.
# For more information, see https://bugzilla.mindrot.org/show_bug.cgi?id=2246
# In the future, if these enhancements are accepted by the upstream in a
# later release, we will remove this patch when we upgrade to that release.
#
--- old/auth-pam.c	2025-03-21 09:29:50.954296050 +0100
+++ new/auth-pam.c	2025-03-21 09:29:51.089387275 +0100
@@ -618,6 +618,66 @@ sshpam_null_conv(int n, sshpam_const str
 
 static struct pam_conv null_conv = { sshpam_null_conv, NULL };
 
+#ifdef PAM_ENHANCEMENT
+char *
+derive_pam_service_name(Authctxt *authctxt)
+{
+	char *svcname = xmalloc(BUFSIZ);
+
+	/*
+	 * If PamServiceName is set we use that for everything, including
+	 * SSHv1
+	 */
+	if (options.pam_service_name != NULL) {
+		(void) strlcpy(svcname, options.pam_service_name, BUFSIZ);
+		return (svcname);
+	}
+
+	char *method_name = authctxt->authmethod_name;
+
+	if (!method_name)
+		fatal("Userauth method unknown while starting PAM");
+
+	/*
+	 * For SSHv2 we use "sshd-<userauth name>
+	 * The "sshd" prefix can be changed via the PAMServicePrefix
+	 * sshd_config option.
+	 */
+	if (strcmp(method_name, "none") == 0) {
+		snprintf(svcname, BUFSIZ, "%s-none",
+		    options.pam_service_prefix);
+	} else if (strcmp(method_name, "password") == 0) {
+		snprintf(svcname, BUFSIZ, "%s-password",
+		    options.pam_service_prefix);
+	} else if (strcmp(method_name, "keyboard-interactive") == 0) {
+		/* "keyboard-interactive" is too long, shorten it */
+		snprintf(svcname, BUFSIZ, "%s-kbdint",
+		    options.pam_service_prefix);
+	} else if ((strcmp(method_name, "publickey") == 0) ||
+	    (strcmp(method_name, "publickey-hostbound-v00@openssh.com") == 0)) {
+		/* "publickey" is too long, shorten it */
+		snprintf(svcname, BUFSIZ, "%s-pubkey",
+		    options.pam_service_prefix);
+	} else if (strcmp(method_name, "hostbased") == 0) {
+		snprintf(svcname, BUFSIZ, "%s-hostbased",
+		    options.pam_service_prefix);
+	} else if (strncmp(method_name, "gssapi-", 7) == 0) {
+		/*
+		 * Although OpenSSH only supports "gssapi-with-mic"
+		 * for now. We will still map any userauth method
+		 * prefixed with "gssapi-" to the gssapi PAM service.
+		 */ 
+		snprintf(svcname, BUFSIZ, "%s-gssapi",
+		    options.pam_service_prefix);
+	} else {
+		/* Unknown userauth method */
+		fatal_f("Unknown userauth method %s", method_name);
+	}
+
+	return svcname;
+}
+#endif /* PAM_ENHANCEMENT */
+
 static int
 sshpam_store_conv(int n, sshpam_const struct pam_message **msg,
     struct pam_response **resp, void *data)
@@ -690,8 +750,19 @@ sshpam_init(struct ssh *ssh, Authctxt *a
 	const char **ptr_pam_user = &pam_user;
 	int r;
 
+#ifdef PAM_ENHANCEMENT
+	const char *pam_service;
+	const char **ptr_pam_service = &pam_service;
+	char *svc = NULL;
+
+	svc = derive_pam_service_name(authctxt);
+	debug3("PAM service is %s", svc);
+#else /* PAM_ENHANCEMENT */
+
 	if (options.pam_service_name == NULL)
 		fatal_f("internal error: NULL PAM service name");
+#endif /* PAM_ENHANCEMENT */
+
 #if defined(PAM_SUN_CODEBASE) && defined(PAM_MAX_RESP_SIZE)
 	/* Protect buggy PAM implementations from excessively long usernames */
 	if (strlen(user) >= PAM_MAX_RESP_SIZE)
@@ -705,18 +776,61 @@ sshpam_init(struct ssh *ssh, Authctxt *a
 		}
 	}
 	if (sshpam_handle != NULL) {
+#ifdef PAM_ENHANCEMENT
+		/* get the pam service name */
+		sshpam_err = pam_get_item(sshpam_handle, 
+		    PAM_SERVICE, (sshpam_const void **)ptr_pam_service);
+		if (sshpam_err != PAM_SUCCESS) 
+			fatal("Failed to get the PAM service name"); 
+		debug3("Previous pam_service is %s", pam_service ?
+		    pam_service : "NULL");
+      
+		/* get the pam user name */
+		sshpam_err = pam_get_item(sshpam_handle,
+		    PAM_USER, (sshpam_const void **)ptr_pam_user);
+		
+		/*  
+		 * only need to re-start if either user or service is
+		 * different.
+		 */ 
+		if (sshpam_err == PAM_SUCCESS && strcmp(user, pam_user) == 0
+		    && strncmp(svc, pam_service, strlen(svc)) == 0) {
+			free(svc);
+			return (0); 
+		}
+		 
+		/*
+		 * Clean up previous PAM state.  No need to clean up session
+		 * and creds.
+		 */
+		sshpam_authenticated = 0; 
+		sshpam_account_status = -1;
+
+		sshpam_err = pam_set_item(sshpam_handle, PAM_CONV, NULL);
+		if (sshpam_err != PAM_SUCCESS)
+			debug3("Cannot remove PAM conv"); /* a warning only */
+#else /* Original */
 		/* We already have a PAM context; check if the user matches */
 		sshpam_err = pam_get_item(sshpam_handle,
 		    PAM_USER, (sshpam_const void **)ptr_pam_user);
 		if (sshpam_err == PAM_SUCCESS && strcmp(user, pam_user) == 0)
 			return (0);
+#endif /* PAM_ENHANCEMENT */
 		pam_end(sshpam_handle, sshpam_err);
 		sshpam_handle = NULL;
 	}
 	debug("PAM: initializing for \"%s\" with service \"%s\"", user,
 	    options.pam_service_name);
+
+#ifdef PAM_ENHANCEMENT
+        debug3("Starting PAM service %s for user %s method %s", svc, user,
+            authctxt->authmethod_name);
+        sshpam_err = pam_start(svc, user, &store_conv, &sshpam_handle);
+        free(svc);
+#else /* Original */
 	sshpam_err = pam_start(options.pam_service_name, user,
 	    &store_conv, &sshpam_handle);
+#endif /* PAM_ENHANCEMENT */
 	sshpam_authctxt = authctxt;
 
 	if (sshpam_err != PAM_SUCCESS) {
--- old/auth.h	2025-02-18 09:15:08.000000000 +0100
+++ new/auth.h	2025-03-21 09:29:51.089797915 +0100
@@ -95,6 +95,9 @@ struct Authctxt {
 
 	/* Information exposed to session */
 	struct sshbuf	*session_info;	/* Auth info for environment */
+#ifdef PAM_ENHANCEMENT
+	char		*authmethod_name;
+#endif 
 };
 
 /*
--- old/auth2.c	2025-02-18 09:15:08.000000000 +0100
+++ new/auth2.c	2025-03-21 09:29:51.090396880 +0100
@@ -305,7 +305,15 @@ input_userauth_request(int type, u_int32
 #endif
 		}
 #ifdef USE_PAM
+#ifdef PAM_ENHANCEMENT
+		/*
+		 * Start PAM here and once only, if each userauth does not
+		 * has its own PAM service.
+		 */
+		if (options.use_pam && !options.pam_service_per_authmethod)
+#else
 		if (options.use_pam)
+#endif /* PAM_ENHANCEMENT */
 			mm_start_pam(ssh);
 #endif
 		ssh_packet_set_log_preamble(ssh, "%suser %s",
@@ -342,6 +350,16 @@ input_userauth_request(int type, u_int32
 	/* try to authenticate user */
 	m = authmethod_lookup(authctxt, method);
 	if (m != NULL && authctxt->failures < options.max_authtries) {
+#if defined(USE_PAM) && defined(PAM_ENHANCEMENT)
+		/* start PAM service for each userauth */
+		if (options.use_pam && options.pam_service_per_authmethod) {
+			if (authctxt->authmethod_name != NULL)
+				free(authctxt->authmethod_name);
+			authctxt->authmethod_name = xstrdup(method);
+			mm_inform_authmethod(method);
+			mm_start_pam(ssh);
+		}
+#endif
 		debug2("input_userauth_request: try method %s", method);
 		authenticated =	m->userauth(ssh, method);
 	}
@@ -367,6 +385,10 @@ userauth_finish(struct ssh *ssh, int aut
 	char *methods;
 	int r, partial = 0;
 
+#ifdef  PAM_ENHANCEMENT
+	debug3("%s: entering", __func__);
+#endif
+
 	if (authenticated) {
 		if (!authctxt->valid) {
 			fatal("INTERNAL ERROR: authenticated invalid user %s",
@@ -390,6 +412,30 @@ userauth_finish(struct ssh *ssh, int aut
 	}
 
 	if (authenticated && options.num_auth_methods != 0) {
+
+#if defined(USE_PAM) && defined(PAM_ENHANCEMENT)
+		/*
+		 * If each userauth has its own PAM service, then PAM need to
+		 * perform account check for this service.
+		 */
+		if (options.use_pam && options.pam_service_per_authmethod &&
+		    !mm_do_pam_account()) {
+			/* if PAM returned a message, send it to the user */
+			if (sshbuf_len(loginmsg) > 0) {
+				if ((r = sshbuf_put(loginmsg, "\0", 1)) != 0)
+					fatal("%s: buffer error: %s",
+					    __func__, ssh_err(r));
+				userauth_send_banner(ssh, sshbuf_ptr(loginmsg));
+				if ((r = ssh_packet_write_wait(ssh)) != 0) {
+					sshpkt_fatal(ssh, r,
+					    "%s: send PAM banner", __func__);
+				}
+			}
+
+			fatal("Access denied for user %s by PAM account "
+			    "configuration", authctxt->user);
+		}
+#endif
 		if (!auth2_update_methods_lists(authctxt, method, submethod)) {
 			authenticated = 0;
 			partial = 1;
@@ -407,7 +453,19 @@ userauth_finish(struct ssh *ssh, int aut
 		return;
 
 #ifdef USE_PAM
+#ifdef PAM_ENHANCEMENT
+	/*              
+	 * PAM needs to perform account checks after auth. However, if each
+	 * userauth has its own PAM service and options.num_auth_methods != 0,
+	 * then no need to perform account checking, because it was done
+	 * already.
+	 */
+	if (options.use_pam && authenticated &&
+	    !(options.num_auth_methods != 0 &&
+	    options.pam_service_per_authmethod)){
+#else
 	if (options.use_pam && authenticated) {
+#endif /* PAM_ENHANCEMENT */
 		int r, success = mm_do_pam_account();
 
 		/* If PAM returned a message, send it to the user. */
--- old/monitor.c	2025-02-18 09:15:08.000000000 +0100
+++ new/monitor.c	2025-03-21 09:29:51.091367695 +0100
@@ -118,6 +118,9 @@ int mm_answer_sign(struct ssh *, int, st
 int mm_answer_pwnamallow(struct ssh *, int, struct sshbuf *);
 int mm_answer_auth2_read_banner(struct ssh *, int, struct sshbuf *);
 int mm_answer_authserv(struct ssh *, int, struct sshbuf *);
+#ifdef PAM_ENHANCEMENT
+int mm_answer_authmethod(struct ssh*, int, struct sshbuf *);
+#endif
 int mm_answer_authpassword(struct ssh *, int, struct sshbuf *);
 int mm_answer_bsdauthquery(struct ssh *, int, struct sshbuf *);
 int mm_answer_bsdauthrespond(struct ssh *, int, struct sshbuf *);
@@ -190,10 +193,17 @@ struct mon_table mon_dispatch_proto20[]
     {MONITOR_REQ_SIGN, MON_ONCE, mm_answer_sign},
     {MONITOR_REQ_PWNAM, MON_ONCE, mm_answer_pwnamallow},
     {MONITOR_REQ_AUTHSERV, MON_ONCE, mm_answer_authserv},
+#ifdef PAM_ENHANCEMENT
+    {MONITOR_REQ_AUTHMETHOD, MON_ISAUTH, mm_answer_authmethod},
+#endif
     {MONITOR_REQ_AUTH2_READ_BANNER, MON_ONCE, mm_answer_auth2_read_banner},
     {MONITOR_REQ_AUTHPASSWORD, MON_AUTH, mm_answer_authpassword},
 #ifdef USE_PAM
+#ifdef PAM_ENHANCEMENT
+    {MONITOR_REQ_PAM_START, MON_ISAUTH, mm_answer_pam_start},
+#else
     {MONITOR_REQ_PAM_START, MON_ONCE, mm_answer_pam_start},
+#endif /* PAM_ENHANCEMENT */
     {MONITOR_REQ_PAM_ACCOUNT, 0, mm_answer_pam_account},
     {MONITOR_REQ_PAM_INIT_CTX, MON_ONCE, mm_answer_pam_init_ctx},
     {MONITOR_REQ_PAM_QUERY, 0, mm_answer_pam_query},
@@ -304,6 +314,25 @@ monitor_child_preauth(struct ssh *ssh, s
 
 		/* Special handling for multiple required authentications */
 		if (options.num_auth_methods != 0) {
+#if defined(USE_PAM) && defined(PAM_ENHANCEMENT)
+			/* 
+			 * If each userauth has its own PAM service, then PAM
+			 * need to perform account check for this service.
+			 */
+			if (options.use_pam && authenticated &&
+			    options.pam_service_per_authmethod) {
+				struct sshbuf *m;
+
+				if ((m = sshbuf_new()) == NULL)
+					fatal("%s: sshbuf_new failed",
+					    __func__);
+				mm_request_receive_expect(pmonitor->m_sendfd,
+				    MONITOR_REQ_PAM_ACCOUNT, m);
+				authenticated = mm_answer_pam_account(
+				    ssh, pmonitor->m_sendfd, m);
+				sshbuf_free(m);
+			}
+#endif
 			if (authenticated &&
 			    !auth2_update_methods_lists(authctxt,
 			    auth_method, auth_submethod)) {
@@ -321,8 +350,21 @@ monitor_child_preauth(struct ssh *ssh, s
 			    !auth_root_allowed(ssh, auth_method))
 				authenticated = 0;
 #ifdef USE_PAM
-			/* PAM needs to perform account checks after auth */
+#ifdef PAM_ENHANCEMENT
+			/*
+			 * PAM needs to perform account checks after auth.
+			 * However, if each userauth has its own PAM service
+			 * and options.num_auth_methods != 0, then no need to
+			 * perform account checking, because it was done 
+			 * already.
+			 */
+			if (options.use_pam && authenticated &&
+			    !(options.num_auth_methods != 0 &&
+			    options.pam_service_per_authmethod)) {
+#else
+ 			/* PAM needs to perform account checks after auth */
 			if (options.use_pam && authenticated) {
+#endif /* PAM_ENHANCEMENT */
 				struct sshbuf *m;
 
 				if ((m = sshbuf_new()) == NULL)
@@ -826,6 +868,10 @@ mm_answer_pwnamallow(struct ssh *ssh, in
 	/* Allow service/style information on the auth context */
 	monitor_permit(mon_dispatch, MONITOR_REQ_AUTHSERV, 1);
 	monitor_permit(mon_dispatch, MONITOR_REQ_AUTH2_READ_BANNER, 1);
+#ifdef PAM_ENHANCEMENT
+	/* Allow authmethod information on the auth context */
+	monitor_permit(mon_dispatch, MONITOR_REQ_AUTHMETHOD, 1);
+#endif
 
 #ifdef USE_PAM
 	if (options.use_pam)
@@ -850,6 +896,27 @@ int mm_answer_auth2_read_banner(struct s
 	return (0);
 }
 
+#ifdef PAM_ENHANCEMENT
+int
+mm_answer_authmethod(struct ssh *ssh, int sock, struct sshbuf *m)
+{
+	int r;
+
+	monitor_permit_authentications(1);
+
+	if ((r = sshbuf_get_cstring(m, &authctxt->authmethod_name, NULL)) != 0)
+		fatal("%s: buffer error: %s", __func__, ssh_err(r));
+	debug3("%s: authmethod_name=%s", __func__, authctxt->authmethod_name);
+
+	if (strlen(authctxt->authmethod_name) == 0) {
+		free(authctxt->authmethod_name);
+		authctxt->authmethod_name = NULL;
+	}
+
+	return (0);
+}
+#endif /* PAM_ENHANCEMENT */
+
 int
 mm_answer_authserv(struct ssh *ssh, int sock, struct sshbuf *m)
 {
--- old/monitor.h	2025-02-18 09:15:08.000000000 +0100
+++ new/monitor.h	2025-03-21 09:29:51.091673815 +0100
@@ -63,6 +63,9 @@ enum monitor_reqtype {
 	MONITOR_REQ_PAM_FREE_CTX = 110, MONITOR_ANS_PAM_FREE_CTX = 111,
 	MONITOR_REQ_AUDIT_EVENT = 112, MONITOR_REQ_AUDIT_COMMAND = 113,
 
+#ifdef PAM_ENHANCEMENT
+	MONITOR_REQ_AUTHMETHOD = 114,
+#endif 
 };
 
 struct ssh;
--- old/monitor_wrap.c	2025-02-18 09:15:08.000000000 +0100
+++ new/monitor_wrap.c	2025-03-21 09:29:51.092170255 +0100
@@ -452,6 +452,27 @@ mm_inform_authserv(char *service, char *
 	sshbuf_free(m);
 }
 
+#ifdef PAM_ENHANCEMENT
+/* Inform the privileged process about the authentication method */
+void
+mm_inform_authmethod(char *authmethod)
+{
+	struct sshbuf *m;
+	int r;
+
+	debug3("%s entering", __func__);
+
+	if ((m = sshbuf_new()) == NULL)
+		fatal("%s: sshbuf_new failed", __func__);
+	if ((r = sshbuf_put_cstring(m, authmethod)) != 0)
+		fatal("%s: buffer error: %s", __func__, ssh_err(r));
+
+	mm_request_send(pmonitor->m_recvfd, MONITOR_REQ_AUTHMETHOD, m);
+
+	sshbuf_free(m);
+}
+#endif /* PAM_ENHANCEMENT */
+
 /* Do the password authentication */
 int
 mm_auth_password(struct ssh *ssh, char *password)
--- old/monitor_wrap.h	2025-02-18 09:15:08.000000000 +0100
+++ new/monitor_wrap.h	2025-03-21 09:29:51.092472355 +0100
@@ -45,6 +45,7 @@ DH *mm_choose_dh(int, int, int);
 int mm_sshkey_sign(struct ssh *, struct sshkey *, u_char **, size_t *,
     const u_char *, size_t, const char *, const char *,
     const char *, u_int compat);
+void mm_inform_authmethod(char *);
 void mm_inform_authserv(char *, char *);
 struct passwd *mm_getpwnamallow(struct ssh *, const char *);
 char *mm_auth2_read_banner(void);
--- old/servconf.c	2025-03-21 09:29:50.257414235 +0100
+++ new/servconf.c	2025-03-21 09:33:11.426313825 +0100
@@ -93,6 +93,17 @@ initialize_server_options(ServerOptions
 	/* Portable-specific options */
 	options->use_pam = -1;
 	options->pam_service_name = NULL;
+#ifdef PAM_ENHANCEMENT
+	options->pam_service_prefix = NULL;
+
+	/* 
+	 * Each user method will have its own PAM service by default.
+	 * However, if PAMServiceName is specified or the protocol version
+	 * is not compat20, then there will be only one PAM service for the
+	 * entire user authentication.
+	 */
+	options->pam_service_per_authmethod = 1;
+#endif /* PAM_ENHANCEMENT */
 
 	/* Standard Options */
 	options->num_ports = 0;
@@ -304,8 +315,14 @@ fill_default_server_options(ServerOption
 		options->use_pam = 0;
 #endif
 
+#ifdef PAM_ENHANCEMENT
+	if (options->pam_service_prefix == NULL &&
+	    options->pam_service_name == NULL)
+		options->pam_service_prefix = xstrdup(SSHD_PAM_SERVICE);
+#else /* PAM_ENHANCEMENT */
 	if (options->pam_service_name == NULL)
 		options->pam_service_name = xstrdup(SSHD_PAM_SERVICE);
+#endif /* PAM_ENHANCEMENT */
 
 	/* Standard Options */
 	if (options->num_host_key_files == 0) {
@@ -574,6 +591,9 @@ typedef enum {
 	sMatch, sPermitOpen, sPermitListen, sForceCommand, sChrootDirectory,
 	sUsePrivilegeSeparation, sAllowAgentForwarding,
 	sHostCertificate, sInclude,
+#ifdef PAM_ENHANCEMENT
+	sPAMServicePrefix,
+#endif
 	sRevokedKeys, sTrustedUserCAKeys, sAuthorizedPrincipalsFile,
 	sAuthorizedPrincipalsCommand, sAuthorizedPrincipalsCommandUser,
 	sKexAlgorithms, sCASignatureAlgorithms, sIPQoS, sVersionAddendum,
@@ -726,6 +746,9 @@ static struct {
 	{ "forcecommand", sForceCommand, SSHCFG_ALL },
 	{ "chrootdirectory", sChrootDirectory, SSHCFG_ALL },
 	{ "hostcertificate", sHostCertificate, SSHCFG_GLOBAL },
+#ifdef PAM_ENHANCEMENT
+	{ "pamserviceprefix", sPAMServicePrefix, SSHCFG_GLOBAL },
+#endif
 	{ "revokedkeys", sRevokedKeys, SSHCFG_ALL },
 	{ "trustedusercakeys", sTrustedUserCAKeys, SSHCFG_ALL },
 	{ "authorizedprincipalsfile", sAuthorizedPrincipalsFile, SSHCFG_ALL },
@@ -1397,10 +1420,41 @@ process_server_config_line_depth(ServerO
 			fatal("%s line %d: missing argument.",
 			    filename, linenum);
 		}
+#ifdef PAM_ENHANCEMENT
+		if (options->pam_service_prefix != NULL)
+			fatal("%s line %d: PAMServiceName and PAMServicePrefix"
+			    " are mutually exclusive.", filename, linenum);
+		if (*activep && *charptr == NULL) {
+			*charptr = xstrdup(arg);
+
+			/*
+			* When this option is specified, we will not have
+			* PAM service for each auth method.
+			*/
+			options->pam_service_per_authmethod = 0;
+		}
+#else
+		if (*activep && *charptr == NULL)
+			*charptr = xstrdup(arg);
+#endif /* PAM_ENHANCEMENT */
+		break;
+
+#ifdef PAM_ENHANCEMENT
+	case sPAMServicePrefix:
+		charptr = &options->pam_service_prefix;
+		arg = argv_next(&ac, &av);
+		if (!arg || *arg == '\0')
+			fatal("%s line %d: Missing argument.",
+			    filename, linenum);
+		if (options->pam_service_name != NULL)
+			fatal("%s line %d: PAMServiceName and PAMServicePrefix"
+			    " are mutually exclusive.", filename, linenum);
 		if (*activep && *charptr == NULL)
 			*charptr = xstrdup(arg);
 		break;
 
+#endif /* PAM_ENHANCEMENT */
+
 	/* Standard Options */
 	case sBadOption:
 		goto out;
@@ -3216,8 +3270,18 @@ dump_config(ServerOptions *o)
 	/* integer arguments */
 #ifdef USE_PAM
 	dump_cfg_fmtint(sUsePAM, o->use_pam);
+#ifdef PAM_ENHANCEMENT
+	/* Print at most one and not if they are the default setting. */
+	if (o->pam_service_name != NULL) {
+		dump_cfg_string(sPAMServiceName, o->pam_service_name);
+	} else if (o->pam_service_prefix != NULL &&
+	    strcmp(o->pam_service_prefix, SSHD_PAM_SERVICE) != 0) {
+		dump_cfg_string(sPAMServicePrefix, o->pam_service_prefix);
+	}
+#else
 	dump_cfg_string(sPAMServiceName, o->pam_service_name);
 #endif
+#endif
 	dump_cfg_int(sLoginGraceTime, o->login_grace_time);
 	dump_cfg_int(sX11DisplayOffset, o->x11_display_offset);
 	dump_cfg_int(sMaxAuthTries, o->max_authtries);
--- old/servconf.h	2025-02-18 09:15:08.000000000 +0100
+++ new/servconf.h	2025-03-21 09:29:51.094069840 +0100
@@ -212,6 +212,10 @@ typedef struct {
 
 	int	use_pam;		/* Enable auth via PAM */
 	char   *pam_service_name;
+#ifdef PAM_ENHANCEMENT
+	char	*pam_service_prefix;
+	int	pam_service_per_authmethod;
+#endif
 
 	int	permit_tun;
 
@@ -300,6 +304,7 @@ TAILQ_HEAD(include_list, include_item);
 		M_CP_STROPT(routing_domain); \
 		M_CP_STROPT(permit_user_env_allowlist); \
 		M_CP_STROPT(pam_service_name); \
+		M_CP_STROPT(pam_service_prefix); \
 		M_CP_STRARRAYOPT(authorized_keys_files, num_authkeys_files); \
 		M_CP_STRARRAYOPT(allow_users, num_allow_users); \
 		M_CP_STRARRAYOPT(deny_users, num_deny_users); \
--- old/sshd.8	2025-02-18 09:15:08.000000000 +0100
+++ new/sshd.8	2025-03-21 09:29:51.094545430 +0100
@@ -1018,6 +1018,33 @@ concurrently for different ports, this c
 started last).
 The content of this file is not sensitive; it can be world-readable.
 .El
+
+.Sh SECURITY
+sshd uses pam(3PAM) for password and keyboard-interactive methods as well as 
+for account management, session management, and the password management for all
+authentication methods.
+.Pp
+Each SSHv2 userauth type has its own PAM service name:
+
+.Bd -literal -offset 3n
+
+-----------------------------------------------
+| SSHv2 Userauth       | PAM Service Name     |
+-----------------------------------------------
+| none                 | sshd-none            |
+-----------------------------------------------
+| password             | sshd-password        |
+-----------------------------------------------
+| keyboard-interactive | sshd-kbdint          |
+-----------------------------------------------
+| pubkey               | sshd-pubkey          |
+-----------------------------------------------
+| hostbased            | sshd-hostbased       |
+-----------------------------------------------
+| gssapi-with-mic      | sshd-gssapi          |
+-----------------------------------------------
+.Ed
+
 .Sh SEE ALSO
 .Xr scp 1 ,
 .Xr sftp 1 ,
--- old/sshd_config.5	2025-03-21 09:29:50.391451320 +0100
+++ new/sshd_config.5	2025-03-21 09:29:51.095261530 +0100
@@ -1315,6 +1315,7 @@ Available keywords are
 .Cm MaxAuthTries ,
 .Cm MaxSessions ,
 .Cm PAMServiceName ,
+.Cm PAMServicePrefix ,
 .Cm PasswordAuthentication ,
 .Cm PermitEmptyPasswords ,
 .Cm PermitListen ,
@@ -1383,12 +1384,20 @@ key exchange methods.
 The default is
 .Pa /etc/moduli .
 .It Cm PAMServiceName
-Specifies the service name used for Pluggable Authentication Modules (PAM)
-authentication, authorisation and session controls when
-.Cm UsePAM
-is enabled.
-The default is
-.Cm sshd .
+Specifies the PAM service name for the PAM session. The PAMServiceName and
+PAMServicePrefix options are mutually exclusive and if both set, sshd does not
+start. If this option is set the service name is the same for all user
+authentication methods. The option has no default value. See PAMServicePrefix
+for more information.
+.It Cm PAMServicePrefix
+Specifies the PAM service name prefix for service names used for individual
+user authentication methods. The default is sshd. The PAMServiceName and
+PAMServicePrefix options are mutually exclusive and if both set, sshd does not
+start.
+.Pp 
+For example, if this option is set to admincli, the service name for the
+keyboard-interactive authentication method is admincli-kbdint instead of the
+default sshd-kbdint.
 .It Cm PasswordAuthentication
 Specifies whether password authentication is allowed.
 The default is
@@ -2036,8 +2045,7 @@ If
 is enabled, you will not be able to run
 .Xr sshd 8
 as a non-root user.
-The default is
-.Cm no .
+On Solaris, this option is always enabled.
 .It Cm VersionAddendum
 Optionally specifies additional text to append to the SSH protocol banner
 sent by the server upon connection.
